lib/api/src/modules/refs.ts

```diff
@@ -209,18 +212,28 @@ export const init: ModuleFn = ({ store, provider, singleStory }, { runCheck = tr
       return refs;
     },
 
-    setRef: (id, { stories, ...rest }, ready = false) => {
+    setRef: (id, { stories, v, ...rest }, ready = false) => {
       if (singleStory) return;
       const { storyMapper = defaultStoryMapper } = provider.getConfig();
       const ref = api.getRefs()[id];
-      const after = stories
-        ? addRefIds(
-            transformStoriesRawToStoriesHash(map(stories, ref, { storyMapper }), { provider }),
-            ref
-          )
-        : undefined;
-
-      api.updateRef(id, { stories: after, ...rest, ready });
+
+      let storiesHash: StoriesHash;
+
+      if (stories) {
+        if (v === 2) {
+          storiesHash = transformStoriesRawToStoriesHash(map(stories, ref, { storyMapper }), {
+            provider,
+          });
+        } else if (v === 3) {
+          const index = (stories as unknown) as Record<StoryId, StoryIndexStory>;
+          storiesHash = transformStoryIndexToStoriesHash({ v, stories: index }, { provider });
+        } else {
+          throw new Error(`Composition: Unknown stories.json version: ${v}`);
```

Comment:
Hmm, I wonder if we'll regret this if we ever want to make `v:4` backwards compatible

Reply:
Good catch. Updated!